private/buf/cmd/buf/command/registry/registrylogin/registrylogin.go

```diff
@@ -211,11 +236,102 @@ func inner(
 	}
 	loggedInMessage := fmt.Sprintf("Logged in as %s. Credentials saved to %s.\n", user.Username, netrcFilePath)
 	// Unless we did not prompt at all, print a newline first
-	if !flags.TokenStdin {
+	if !flags.TokenStdin || !flags.NoBrowser {
 		loggedInMessage = "\n" + loggedInMessage
 	}
 	if _, err := container.Stdout().Write([]byte(loggedInMessage)); err != nil {
 		return err
 	}
 	return nil
 }
+
+// doBrowserLogin performs the device authorization grant flow via the browser.
+func doBrowserLogin(
+	ctx context.Context,
+	container appext.Container,
+	remote string,
+) (string, error) {
+	baseURL := "https://" + remote
+	clientName, err := getClientName()
+	if err != nil {
+		return "", err
+	}
+	externalConfig := bufapp.ExternalConfig{}
+	if err := appext.ReadConfig(container, &externalConfig); err != nil {
+		return "", err
+	}
+	appConfig, err := bufapp.NewConfig(container, externalConfig)
+	if err != nil {
+		return "", err
+	}
+	client := httpclient.NewClient(appConfig.TLS)
+	oauth2Client := oauth2.NewClient(baseURL, client)
+	// Register the device.
+	deviceRegistration, err := oauth2Client.RegisterDevice(ctx, &oauth2.DeviceRegistrationRequest{
```

Comment:
Could inspect the error case here to fallback to the `--no-browser` flow for a 501 not implemented error. This would allow pointing at older BSR instances that don't have the required auth endpoints.

Reply:
I like this idea -- I think not breaking for older BSR versions is pretty nice. I wonder if we should have a note about this in the `--help` string, something along the lines of "for old versions of the Buf Schema Registry (BSR), buf registry login defaults to --no-browser".

Reply:
Theres `errors.ErrUnsupported` which seems nice to trigger on content type failures, or any other unsupported response types.